nestkernel/eprop_archiving_node.cpp

```diff
@@ -0,0 +1,253 @@
+{
+  eprop_parameter_map_[ parameter_name ] = parameter_value;
+}
+
+void
+nest::EpropArchivingNode::write_surrogate_gradient_to_history( long time_step, double surrogate_gradient )
+{
+  eprop_history_.push_back( HistEntryEpropArchive( time_step, surrogate_gradient, 0.0 ) );
+}
+
+void
+nest::EpropArchivingNode::write_error_signal_to_history( long time_step, double error_signal )
+{
+  long shift = delay_out_norm_;
+  eprop_history_.push_back( HistEntryEpropArchive( time_step - shift, 0.0, error_signal ) );
+}
+
+void
+nest::EpropArchivingNode::write_learning_signal_to_history( long& time_step,
+  long& delay_out_rec,
+  double& weight,
+  double& error_signal )
+{
+  long shift = delay_rec_out_ + delay_out_norm_ + delay_out_rec;
+
+  std::deque< HistEntryEpropArchive >::iterator it_hist;
+  std::deque< HistEntryEpropArchive >::iterator it_hist_end;
+
+  get_eprop_history( time_step - shift, &it_hist );
+  get_eprop_history( time_step - shift + delay_out_rec, &it_hist_end );
+
+  if ( it_hist != it_hist_end )
+  {
+    it_hist->learning_signal_ += weight * error_signal;
+    ++it_hist;
+  }
+}
+
+void
+nest::EpropArchivingNode::write_firing_rate_reg_to_history( long t_current_update, double f_target, double c_reg )
+{
+  long const update_interval = kernel().simulation_manager.get_eprop_update_interval().get_steps();
+  long shift = Time::get_resolution().get_steps();
+
+  double f_av = n_spikes_ / static_cast< double >( update_interval );
+  double f_target_ = f_target / 1000.0; // convert to kHz
+  double firing_rate_reg = c_reg * ( f_av - f_target_ ) / static_cast< double >( update_interval );
+
+  firing_rate_reg_history_.push_back( HistEntryEpropFiringRateReg( t_current_update + shift, firing_rate_reg ) );
+}
+
+const long
+nest::EpropArchivingNode::get_shift() const
+{
+  long shift_rec = offset_gen_ + delay_in_rec_;
+  long shift_out = offset_gen_ + delay_in_rec_ + delay_rec_out_;
+
+  return get_name() == "eprop_readout" ? shift_out : shift_rec;
+}
+
+const long
+nest::EpropArchivingNode::get_delay_out_norm() const
+{
+  return delay_out_norm_;
+}
+
+std::map< std::string, double >&
+nest::EpropArchivingNode::get_eprop_parameter_map()
+{
+  return eprop_parameter_map_;
+}
+
+void
+nest::EpropArchivingNode::get_eprop_history( long time_step, std::deque< HistEntryEpropArchive >::iterator* it )
+{
+  if ( eprop_history_.empty() )
+  {
+    *it = eprop_history_.end();
+    return;
+  }
+
+  *it = std::lower_bound( eprop_history_.begin(), eprop_history_.end(), time_step );
+}
+
+double
+nest::EpropArchivingNode::get_firing_rate_reg( long time_step )
+{
+  if ( firing_rate_reg_history_.empty() )
+    return 0.0;
+
+  long const update_interval = kernel().simulation_manager.get_eprop_update_interval().get_steps();
+
+  std::vector< HistEntryEpropFiringRateReg >::iterator it;
+
+  it =
+    std::lower_bound( firing_rate_reg_history_.begin(), firing_rate_reg_history_.end(), time_step + update_interval );
+
+  return it->firing_rate_reg_;
```

Comment:
```suggestion
  auto it =
    std::lower_bound( firing_rate_reg_history_.begin(), firing_rate_reg_history_.end(), time_step + update_interval );

  return it->firing_rate_reg_;
```
and maybe a different name for `it`. What point in the history does it represent?